fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py

```diff
@@ -0,0 +1,167 @@
+class CyclicalSGLDPosterior(SGMCMCPosteriorMixin, Posterior):
+    def __init__(
+        self,
+        joint: Joint,
+        posterior_approximator: CyclicalSGLDPosteriorApproximator,
+    ):
+        """
+        Cyclical Stochastic Gradient Langevin Dynamics (SGLD) approximate posterior class.
+
+        Parameters
+        ----------
+        joint: Joint
+            A Joint distribution object.
+        posterior_approximator: CyclicalSGLDPosteriorApproximator
+            A cyclical SGLD posterior approximator.
+        """
+        super().__init__(
+            joint=joint, posterior_approximator=posterior_approximator
+        )
+
+    def __str__(self):
+        return CYCLICAL_SGLD_NAME
+
+    def fit(
+        self,
+        train_data_loader: DataLoader,
+        val_data_loader: Optional[DataLoader] = None,
+        fit_config: FitConfig = FitConfig(),
+        **kwargs,
+    ) -> Status:
+        super()._checks_on_fit_start(fit_config, map_fit_config=None)
+
+        if fit_config.optimizer.method is not None:
+            logging.info(f"`FitOptimizer` method in CyclicalSGLD is ignored.")
+
+        fit_config.optimizer.method = cyclical_sgld_integrator(
+            rng_key=self.rng.get(),
+            init_step_size=self.posterior_approximator.init_step_size,
+            cycle_length=self.posterior_approximator.cycle_length,
+            exploration_ratio=self.posterior_approximator.exploration_ratio,
+            preconditioner=self.posterior_approximator.preconditioner,
+        )
+
+        trainer_cls = select_trainer_given_devices(
+            devices=fit_config.processor.devices,
+            BaseTrainer=MAPTrainer,
+            JittedTrainer=JittedMAPTrainer,
+            MultiDeviceTrainer=MultiDeviceMAPTrainer,
+            disable_jit=fit_config.processor.disable_jit,
+        )
+
+        save_checkpoint_dir = (
+            pathlib.Path(fit_config.checkpointer.save_checkpoint_dir) / "c"
+            if fit_config.checkpointer.save_checkpoint_dir
+            else None
+        )
+        trainer = trainer_cls(
+            predict_fn=self.joint.likelihood.prob_output_layer.predict,
+            save_checkpoint_dir=save_checkpoint_dir,
+            save_every_n_steps=fit_config.checkpointer.save_every_n_steps,
+            keep_top_n_checkpoints=fit_config.checkpointer.keep_top_n_checkpoints,
+            disable_training_metrics_computation=fit_config.monitor.disable_training_metrics_computation,
+            eval_every_n_epochs=fit_config.monitor.eval_every_n_epochs,
+            early_stopping_monitor=fit_config.monitor.early_stopping_monitor,
+            early_stopping_min_delta=fit_config.monitor.early_stopping_min_delta,
+            early_stopping_patience=fit_config.monitor.early_stopping_patience,
+        )
+
+        if super()._is_state_available_somewhere(fit_config):
+            state = self._restore_state_from_somewhere(fit_config=fit_config)
+        else:
+            state = self._init_state(
```

Comment:
Right, I was going to omit preliminary MAP from Cyclical SGLD because it performs SGD steps in the beginning of each cycle, but forgot about the partial freezing / inference on a subset of parameters use case. Will add it back.